### PR TITLE
Fix failing test: pull-kubernetes-e2e-autoscaling-hpa-cm

### LINKED ISSUE #121738: [sig-autoscaling] [HPA] [Feature:CustomMetricsAutoscaling] with multiple metrics of different types should not scale down when one metric is missing
### Failure cluster [342eb6fbb29170f44f82](https://go.k8s.io/triage#342eb6fbb29170f44f82)
https://storage.googleapis.com/k8s-triage/index.html?test=CustomMetricsAutoscaling

**First Failure  in  Nov 02 16:01:48Nov 02 2023, 16:01:48 UTC+0800**

- first failure: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-hpa-cm/1719988161754435584/artifacts/metadata.json -> commit 227d1b2357d93a  Added in #119762
- last success: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-hpa-cm/1719967273617199104/artifacts/metadata.json -> commit 6b8b1881491e26 

- the k/k commit is https://github.com/kubernetes/kubernetes/pull/119762 (not sure if it is related)

/cc @smarterclayton @liggitt  @aojea @AxeZhan 

##### Error text:
```
[FAILED] Failed to get replication controller &Deployment{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:DeploymentSpec{Replicas:nil,Selector:nil,Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {[] [] [] []  <nil> <nil>  map[]   <nil>  false false false <nil> nil []   nil  [] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil> nil <nil> [] []}},Strategy:DeploymentStrategy{Type:,RollingUpdate:nil,},MinReadySeconds:0,RevisionHistoryLimit:nil,Paused:false,ProgressDeadlineSeconds:nil,},Status:DeploymentStatus{ObservedGeneration:0,Replicas:0,UpdatedReplicas:0,AvailableReplicas:0,UnavailableReplicas:0,Conditions:[]DeploymentCondition{},ReadyReplicas:0,CollisionCount:nil,},}: client rate limiter Wait returned an error: context deadline exceeded
In [It] at: test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go:617 @ 11/03/23 20:49:46.812

```
#### Recent failures:
[2023/11/6 04:32:26 ci-kubernetes-e2e-gci-gce-autoscaling-hpa-cm](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-hpa-cm/1721264227454291968)
[2023/11/6 04:32:26 ci-kubernetes-e2e-gci-gce-autoscaling-migs-hpa](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-migs-hpa/1721264227517206528)
[2023/11/6 03:08:26 ci-kubernetes-e2e-gci-gce-autoscaling-migs-hpa](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-migs-hpa/1721243087700758528)
[2023/11/6 03:03:26 ci-kubernetes-e2e-gci-gce-autoscaling-hpa-cm](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-hpa-cm/1721241829648633856)
[2023/11/6 01:43:26 ci-kubernetes-e2e-gci-gce-autoscaling-migs-hpa](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-migs-hpa/1721221696221876224)

/kind failing-test

/sig autoscaling

### PR DESCRIPTION
**Pull request by AxeZhan**
#### What type of PR is this?
/kind failing-test

#### What this PR does / why we need it:
We have some failing e2e tests. 
`ensureDesiredReplicasInRange` is a function created to receive a timeout error(Intentional).
As this timeout is expected, I think we can't throw a panic using framework.Failf, which will case the whole test fail.
Instead we should just return error, and let the check in L631 do it's job.
#### Which issue(s) this PR fixes:

Fixes #121738 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by AxeZhan** [comment]
Hmm, seems still failing blob:https://prow.k8s.io/b3570d5b-c7a9-4561-82e1-460d4a029ea8
```
Dec 22 07:41:53.736: INFO: Failed to get replication controller &Deployment{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:DeploymentSpec{Replicas:nil,Selector:nil,Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {[] [] [] []  <nil> <nil>  map[]   <nil>  false false false <nil> nil []   nil  [] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil> nil <nil> [] []}},Strategy:DeploymentStrategy{Type:,RollingUpdate:nil,},MinReadySeconds:0,RevisionHistoryLimit:nil,Paused:false,ProgressDeadlineSeconds:nil,},Status:DeploymentStatus{ObservedGeneration:0,Replicas:0,UpdatedReplicas:0,AvailableReplicas:0,UnavailableReplicas:0,Conditions:[]DeploymentCondition{},ReadyReplicas:0,CollisionCount:nil,},}: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline
```

File: test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go [hunk 1]
```diff
@@ -611,23 +612,26 @@ func waitForReplicas(ctx context.Context, deploymentName, namespace string, cs c
...
 		} else if replicas > maxDesiredReplicas {
-			return false, fmt.Errorf("number of replicas above target")
+			return false, errAbove
 		} else {
 			return false, nil // Expected number of replicas found. Continue polling until timeout.
 		}
 	})
 	// The call above always returns an error, but if it is timeout, it's OK (condition satisfied all the time).
-	if wait.Interrupted(err) {
+	if err != errBelow && err != errAbove {
```
Comment (aojea) [inline test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go:634]: why changing this? it is only sucesss if this condition matches

Comment (AxeZhan) [reply test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go:634]: I'm actually testing it in this pr, but hope you can help me dig it.

Before I just change the failf to return an error, and the test still fails.
See this comment: https://github.com/kubernetes/kubernetes/pull/122446#issuecomment-1867427179
This error is clearly also a timeout error. However, it seems that it's throw by client-go and will not match `Interrupted` here @aojea .

Comment (AxeZhan) [reply test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go]: It's reproducible https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/122446/pull-kubernetes-e2e-autoscaling-hpa-cm/1738166784302780416
I think when we got a timeout during `cs.AppsV1().Deployments(namespace).Get()`, the timeout returned by client-go doesn't match with `wait.Interrupted`.

Comment (AxeZhan) [reply test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go]: > any other error that is not timeout should fail

I think client-go returned error is a small probability event. And as this function only meant to ensure the number of replicas. It should be ok to ignore that. Also, I can't think of a better way to deal with this error(Wait(n=1) would exceed context deadline) right now.

Comment (aojea) [reply test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go]: if an unexpected error is not expected we must fail, we must be strict on the tests, the end goal of the test is to ensure the code is reliable, is not to just pass, if we relax conditions this translates to unreliable systems

> client rate limiter Wait returned an error: context deadline exceeded

then if this error may be legit, we have to add that error to the condition, but not relax the criteria

File: test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go [hunk 2]
```diff
@@ -627,7 +630,7 @@ func ensureDesiredReplicasInRange(ctx context.Context, deploymentName, namespace
 		}
 	})
 	// The call above always returns an error, but if it is timeout, it's OK (condition satisfied all the time).
-	if wait.Interrupted(err) {
+	if wait.Interrupted(err) || errors.Is(err, fmt.Errorf("rate: Wait(n=%d) would exceed context deadline", 1)) {
```
Comment (sanposhiho) [inline test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go:633]: This `errors.Is` would never be true. Read [the official doc](https://pkg.go.dev/errors) to understand better.
https://go.dev/play/p/L1SuSGuUDCG

If you want to catch such errors with the specific message, `err.Error() == fmt.Sprintf("rate: Wait(n=%d) would exceed context deadline", 1)` is more correct. 
It'd be the best if we could catch error with errors.Is actually though, I'm not sure technical feasibility there, at least it'd take some effort I guess.

```go
package hoge_package

var DeadlineError := errors.New("rate: Wait(n=1) would exceed context deadline")

func WhereThisErrorOriginallyComesFrom(...) error {
  ...
  return DeadlineError
}

===

if wait.Interrupted(err) || errors.Is(err, hoge_package.DeadlineError) {} // It'll work as you expect.
```

**Comment by AxeZhan** [comment]
Seems https://github.com/kubernetes/kubernetes/pull/123473 fixed this, I'll just close this one.

**Pull request closed without merging**